File: packages/web/src/api/learning.ts
import api, { apiUpload } from "./index";

export interface LearningStep {
  index: number;
  name: string;
  description: string;
  required_objects: string[];
  action_type: string;
  timeout_seconds: number;
  is_optional: boolean;
  reference_frame_url: string;
  reference_frame_b64: string;
  reference_frame_timestamp: number;
  ok_criteria: string;
  ng_criteria: string;
  start_sec: number;
  end_sec: number;
  segment_ids: number[];
  review_status: "pending" | "confirmed" | "ignored" | "needs_rework";
  evidence_status: "supported" | "weak" | "missing" | "";
  confirmation_note: string;
  human_reviewed: boolean;
  reviewed_at: string;
  grounding_supported?: boolean | null;
  grounding_confidence?: number | null;
  grounding_issue?: string;
  source_confidence?: number | null;
}

export interface LearningTask {
  id: number;
  task_id: string;
  product_model: string;
  process_name: string;
  video_path: string;
  status: string;
  progress: number;
  steps: LearningStep[];
  analysis_detail: Record<string, unknown>;
  error_message: string;
  template_id: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface LearningTaskListResult {
  items: LearningTask[];
  total: number;
}

export const learningApi = {
  uploadVideo: (formData: FormData, productModel: string, processName: string) =>
    apiUpload.post<{ task_id: string; status: string }>(
      `/learning/upload-video?product_model=${encodeURIComponent(productModel)}&process_name=${encodeURIComponent(processName)}`,
      formData
    ),
  listTasks: (params?: { skip?: number; limit?: number }) =>
    api.get<LearningTaskListResult>("/learning/tasks", { params }),
  getTask: (taskId: string) => api.get<LearningTask>(`/learning/task/${taskId}`),
  updateSteps: (taskId: string, steps: LearningStep[]) =>
    api.put(`/learning/task/${taskId}/steps`, { steps }),
  confirmTask: (taskId: string) =>
    api.post<{ template_id: number; name: string; step_count: number }>(`/learning/task/${taskId}/confirm`),
  retryTask: (taskId: string) => api.post<LearningTask>(`/learning/task/${taskId}/retry`),
  deleteTask: (taskId: string) => api.delete(`/learning/task/${taskId}`),
};

File: packages/server/src/schemas/learning.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, field_validator


class LearningStepItem(BaseModel):
    index: int
    name: str
    description: str = ""
    required_objects: list[str] = []
    action_type: str = ""
    timeout_seconds: int = 30
    is_optional: bool = False
    reference_frame_url: str = ""
    reference_frame_b64: str = ""
    reference_frame_timestamp: float = 0.0
    ok_criteria: str = ""
    ng_criteria: str = ""
    start_sec: float = 0.0
    end_sec: float = 0.0
    segment_ids: list[int] = []
    review_status: str = "pending"
    evidence_status: str = ""
    confirmation_note: str = ""
    human_reviewed: bool = False
    reviewed_at: str = ""
    grounding_supported: bool | None = None
    grounding_confidence: float | None = None
    grounding_issue: str = ""
    source_confidence: float | None = None

    model_config = {"extra": "allow"}


class LearningTaskResponse(BaseModel):
    id: int
    task_id: str
    product_model: str
    process_name: str
    video_path: str
    status: str
    progress: float
    steps: list[dict[str, Any]]
    analysis_detail: dict[str, Any]
    error_message: str
    template_id: int | None
    created_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}

    @field_validator("steps", mode="before")
    @classmethod
    def _steps_none(cls, v: Any) -> Any:
        return v if v is not None else []

    @field_validator("analysis_detail", mode="before")
    @classmethod
    def _detail_none(cls, v: Any) -> Any:
        return v if v is not None else {}

    @field_validator("error_message", mode="before")
    @classmethod
    def _err_none(cls, v: Any) -> Any:
        return v if v is not None else ""

    @field_validator("video_path", mode="before")
    @classmethod
    def _path_none(cls, v: Any) -> Any:
        return v if v is not None else ""

    @field_validator("progress", mode="before")
    @classmethod
    def _progress_none(cls, v: Any) -> Any:
        return 0.0 if v is None else v


class LearningTaskListResponse(BaseModel):
    items: list[LearningTaskResponse]
    total: int


class StepsUpdateRequest(BaseModel):
    steps: list[LearningStepItem]


class TaskCreateResponse(BaseModel):
    task_id: str
    status: str
